### DISCUSSION TITLE
Batching `onDidChangeTextDocument()` events

### DISCUSSION THREAD

**Discussion by RedCMD** [microsoft/vscode-discussions#1606] [Extension Development QnA] [answered]
I've setup `onDidChangeTextDocument()` to parse a document on every change
however parsing can take upwards of 500ms and if the user types too fast
it can very quickly lead to multiple `change` events queuing up
slowing the system to a crawl

https://code.visualstudio.com/api/extension-guides/custom-editors#synchronizing-changes-with-the-textdocument
suggested using `setTimeout()` to debounce the input
which works well
https://davidwalsh.name/javascript-debounce-function

but has one side effect of `symbol outline`/`breadcrumbs` often being out of sync by one batch of change events
(and never resyncing until next input)
I think most people don't notice it because autosave causes a change event 1sec later
allowing everything to resync up
but I don't think that's a very 'clean' way of handling it

problem is `setTimeout()` is affectively async
allowing `onDidChangeTextDocument()` to return, telling VSCode I've 'finished' parsing the document
VSCode then asks for the document symbols asynchronously (while I'm still parsing the document)
and when I'm finished parsing, I have no easy 'clean' way of telling VSCode that the symbols have changed
because `registerDocumentSymbolProvider()` has no `onDidChangeDocumentSymbols`, like what `registerDocumentSemanticTokensProvider` has: `onDidChangeSemanticTokens`

I would like to be able to query how many `change` events are queued up for `onDidChangeTextDocument()`
that way, I can batch parse the document synchronously within `onDidChangeTextDocument()`
removing all timing issues

**Comment by isidorn** [comment]
I do not know. Maybe @bpasero can help

**Comment by bpasero** [comment, marked as answer]
I would suggest to report as feature request.
